### PR TITLE
[RFC] This is a test

### PR DESCRIPTION
**Pull request by glacambre**
This PR adds a selenium testsuite to Tridactyl. Sending input to Tridactyl didn't work a few months ago but now it does, you just have to make sure that you're not sending input too fast (i.e. make sure you send your chars one by one when using sendKeys() and possibly add delays inbetween sendKeys calls).
Don't merge this yet, I think we still need to create a few real tests before deciding whether selenium actually suits our needs (but I think it does).

Help creating new tests and comments are most welcome.

### CODE REVIEW

**Comment by glacambre** [comment]
Installing geckodriver/selenium somehow removed dependencies from package-lock.json. I'll look into it tomorrow.

**Comment by antonva** [comment]
Fantastic that input now works. That's where I stranded every time I tried working with selenium.

I'll write some tests when you fix that package-lock issue.

**Comment by bovine3dom** [comment]
It'd be really nice to test on the Firefox betas as well so we can be warned of any/some upcoming regressions

File: tests/excmds.test.ts [hunk 1]
```diff
@@ -6,6 +6,56 @@ const Options = require("selenium-webdriver/firefox").Options
...
+                return vimToSelenium[modifier] || modifier
+        })
+}
+
+function sendKeys (driver, keys) {
+        return translateKeys(keys).split("")
+                .reduce((prom, key) => {
+                        return prom.then(_ => driver.wait(driver.switchTo().activeElement()))
+                                .then(elem => elem.sendKeys(key))
+                                .then(_ => driver.sleep(5))
```
Comment (bovine3dom) [inline tests/excmds.test.ts]: FWIW, fastest typists in the world type at about 200 WPM peak, so about 1000 CPM, which equates to a 60ms interval between keys. You can therefore safely increase this to 50ms, unless you're concerned about macros.

Comment (glacambre) [reply tests/excmds.test.ts]: I'd rather not make the tests slower that they need to be, running a single test already takes about 7 seconds on my machine :(

**Comment by glacambre** [comment]
Does anybody know why `npm ci` could be failing? It seems to work fine on my machine.

**Comment by glacambre** [comment]
We need to instantiate one selenium driver per test, otherwise tests can share their state and this creates race conditions. Using one driver per test makes things really slow (~25 seconds to run four tests) and thus does not scale.
Unless someone knows how we could work around this state-sharing issue, I think we can deem the selenium approach a dead end and look into mock testing with sinon-chrome.

**Comment by bovine3dom** [comment]
Why do we care about the amount of time it takes to run a test if we're using Travis? If we separate it out into another check we can just ignore it for trivial PRs.

**Comment by glacambre** [comment]
At 6s/test running 600 tests will take an hour. It's not uncommon for testsuites to grow bigger than that (qutebrowser almost has 1000 scenarios for its e2e tests) and I believe Travis has a hard limit of one hour per job, after which the build is killed and marked as failed.

**Comment by bovine3dom** [comment]
What's the bottleneck on the tests? Can we run them in parallel?

Why can't we run the "reload extension" thing that pressing R on web-ext run does?

**Comment by glacambre** [comment]
>What's the bottleneck on the tests?

Driver instantiation.

>Can we run them in parallel?

Yes, they run in parallel if you use a single driver. There seems to be some kind of lock that prevents instantiating selenium drivers in parallel though.

>Why can't we run the "reload extension" thing that pressing R on web-ext run does?

Because this would reload the extension for all of the test running on a same driver (thus probably breaking them) and wouldn't reset Firefox's global state (tabs, cookies, history…).

**Comment by glacambre** [comment]
We can't test `nativeopen()` because geckodriver is ran with `-no-profile` by default and I can't find a way to change than in selenium's docs.

**Comment by bovine3dom** [comment]
Can we separate this out into a separate Travis "check"? I suspect it will be more fragile than the standard "can we build without any errors", and I'd like to be able to test on multiple FF versions at some point.

**Comment by glacambre** [comment]
I moved the E2E tests to another job but it looks like you can't have Travis display multiple checks yet (see https://github.com/travis-ci/travis-ci/issues/5035 ).

**Comment by bovine3dom** [comment]
Circle CI is merged so just waiting on you to rebase this on that :)

**Comment by glacambre** [comment]
Finally got it to work. Can't believe how much time I've wasted on this.

We can't test against firefox stable/beta because `xpinstall.signatures.required` only works on nightly. Fixing this would require signing tridactyl on circleci and I don't have the necessary keys to set that up (if you feel like doing this yourself circleci's docs about secrets is here: https://circleci.com/docs/2.0/config-overview/#secrets-private-keys-tokens-and-scripts ).

**Comment by bovine3dom** [comment]
We'd have to sign it as a 'beta' version right? Seems like a real headache if we're keeping track of multiple PRs.

How does `web-ext run` work, though? That doesn't seem to be nightly, just uses FF66.05 for me. So perhaps we don't need to sign it after all?

**Comment by glacambre** [comment]
>How does web-ext run work, though?

Web-ext seems to use the temporary addon mechanism (which works like loading the addon in `about:debugging`) while selenium uses the "real" install mechanism (which makes the addon appear in `about:addons`).

**Comment by bovine3dom** [comment]
Seems like an obvious thing to fix upstream someday. Is this good to merge, then, or do you want to add more tests? It looks good to me.

**Comment by glacambre** [comment]
Yep, i'd say it's ready.

**Pull request merged**